crates/core/src/binding.rs

```diff
@@ -461,7 +462,9 @@ impl<'a> Binding<'a> {
     pub fn text(&self) -> String {
         match self {
             Binding::Empty(_, _, _) => "".to_string(),
-            Binding::Node(source, node) => node_text(source, node).to_string(),
+            Binding::Node(source, node) => {
+                NodeWithSource::new(node.clone(), source).text().to_string()
+            }
```

Comment:
Not a high priority now, but I think `Binding::Node` can be adjusted to have a `NodeWithSource` instead of separate fields. That'll probably enable a lot of code simplifications down the line.